### PR TITLE
Fix #10 deleted test cases are still present in the tsuite file

### LINKED ISSUE #10: Deleted test cases are still present in the tsuite file
Create a test suite and add two test cases (T1 and T2) to it
Delete test case T2 and generate code.  The following error is displayed:

image

Fix

- Before deleting a test case, make sure it is not in any test suite.   If it is, list the test suites it is in and prompt the user if they wish to remove from all test suite and delete?

### PR DESCRIPTION
**Pull request by vothaison**
#10

### CODE REVIEW

File: electron/worker/actions/generateProjectMetadata.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import { IActionResult } from "./shared";
+
+export const generateProjectMetadata = async function (rmprojFile: IRmProjFile): Promise<IActionResult> {
+  return await new Promise<IActionResult>((rs, _) => {
+    const workerPath = path.join(__dirname, "../worker.js");
+    const worker = new Worker(workerPath);
+
+    worker.on("message", (event: IProgressEvent) => {
+      if (event.type === "finish") {
+        rs({ isSuccess: true, data: event.data, errorMessage: "E" });
```
Comment (quockhanh) [inline electron/worker/actions/generateProjectMetadata.ts:17]: Why `errorMessage: "E"` ?

File: electron/worker/actions/generateProjectMetadata.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import path from "path";
```
Comment (quockhanh) [inline electron/worker/actions/generateProjectMetadata.ts:1]: Rename this file from  `generateProjectMetadata.ts` to `genProjectMetadata.ts`, to match convention with `genCode.ts`

File: electron/IPC/codeGenerator.ts [hunk 3]
```diff
@@ -53,183 +62,200 @@ export default codeGenearator;
...
+    } finally {
+        browserWindow.webContents.send("finish", ipcRs);
+    }
+}
+
+async function generateProjectMetadata(
+    browserWindow: BrowserWindow,
+    event: Electron.IpcMainEvent,
+    projectFile: IRmProjFile
+): Promise<IIpcGenericResponse<ITestSuitesMetadata>> {
```
Comment (quockhanh) [inline electron/IPC/codeGenerator.ts:239]: Rename `ITestSuitesMetadata` (in common) to `IProjectMetadata`.  Because later, this interface does not only contain `suites` info, but also the other

File: electron/IPC/codeGenerator.ts [hunk 4]
```diff
@@ -53,183 +62,200 @@ export default codeGenearator;
...
-    const logFilePath = path.join(logsFolder, logFileName);
-    await fileSystem.writeFile(logFilePath, sb.toString());
-
-    ipcRs.data = { logFile: path.join(StandardFolder.Logs, logFileName) };
-  } finally {
-    browserWindow.webContents.send("finish", ipcRs);
-  }
 }
 
 async function getSuitesMetadata(
```
Comment (quockhanh) [inline electron/IPC/codeGenerator.ts:248]: Rename this method (and other related stuff), from `getSuitesMetadata` to `getProjectMetadata`

File: svelte/src/lib/ipc/codeGenerator.ts [hunk 5]
```diff
@@ -44,6 +44,20 @@ const prerequire = async (rmprojFile: IRmProjFile, options?: { apiKey?: string }
...
+/** Gen metadata of all test suites */
+const generateProjectMetadata = async (
+    projFile: IRmProjFile,
+    options?: { apiKey?: string }
+): Promise<ICodeGenMeta | null> => {
+    const response = await getApi(options?.apiKey).invoke('generateProjectMetadata', projFile);
+    if (response.isSuccess) {
+        return response.data;
+    } else {
+        console.error(`Cannot generate codegen metadata. ${response.errorMessage}`);
```
Comment (quockhanh) [inline svelte/src/lib/ipc/codeGenerator.ts:56]: Should be `Cannot generate **project** metadata`

File: svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  import { StandardFolder, type ITestCase, type ITestSuite } from "rockmelonqa.common/file-defs";
+  import { getContext, onDestroy, onMount } from "svelte";
+
+  /** Toggle on dialog */
+  export let showDialog: boolean = false;
+  /** File system physical path */
+  export let testCaseFilePath: string;
+  /** Path on tree (used for closing coresponding tab after deleting testcase node) */
+  export let nodePath: string;
+  /** Array of relative paths of the test suites that contains the test case being deleted */
```
Comment (quockhanh) [inline svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte:19]: 'Array of test suite relative paths, which contain the test case being deleted'

File: svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    return testCaseId;
+  };
+
+  const getTestSuiteFilePaths = () => {
+    const relatedSuiteFilePaths = relatedTestSuiteRelPaths.map((relPath) => {
+      const suiteFilePath = combinePath(
+        [
+          $appState.projectFile!.folderPath,
+          StandardFolder.TestSuites,
+          relPath.replaceAll("/", uiContext.pathSeparator),
```
Comment (quockhanh) [inline svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte:48]: Dont hardcode "/",   should use   Node.ts->PATH_SEPARATOR   constant instead

File: svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+  const closeTestCaseEditorTab = () => {
+    const tabIndex = $appState.tabs.findIndex((t) => t.id === nodePath);
+    if (tabIndex >= 0) {
+      appStateDispatch({ type: AppActionType.CloseTab, tabIndex: tabIndex });
+    }
+  };
+
+  const handleDeleteClick = async () => {
+    showDialog = false;
```
Comment (quockhanh) [inline svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte:88]: Let close dialog at the end of function. That is coding convention
Sometimes, if the logic throws error, we need to show error message in the dialog, so we may need to keep it in that case (or only close if nothing wrong)

File: svelte/src/lib/components/FileExplorer/Node.svelte [hunk 9]
```diff
@@ -1,295 +1,338 @@
...
+    }
+
+    console.log("generateProjectMetadata", meta);
+    console.log("testCase", testCase);
+    console.log("relatedTestSuiteRelPaths", relatedTestSuiteRelPaths);
+    showDeleteTestCaseDialog = true;
+  };
+
+  const handleMenuDelete = async () => {
+    if (type === FileType.TCase) {
```
Comment (quockhanh) [inline svelte/src/lib/components/FileExplorer/Node.svelte:165]: Important: How about if delete whole  test-case sub-folder?

File: svelte/src/lib/components/FileExplorer/Node.svelte [hunk 10]
```diff
@@ -1,295 +1,338 @@
...
+      relatedTestSuiteRelPaths = meta.testSuites
+        .filter((suiteFile) => suiteFile.content.testcases.some((tcId) => tcId === testCase?.content.id))
+        .map((suiteFile) => {
+          const suiteFileFullPath = combinePath([suiteFile.folderPath, suiteFile.fileName], uiContext.pathSeparator);
+          const suitesFolder = combinePath(
+            [$appState.projectFile!.folderPath, StandardFolder.TestSuites],
+            uiContext.pathSeparator
+          );
+          const suiteFileRelPath = suiteFileFullPath.replace(suitesFolder, "").substring(1);
+          return suiteFileRelPath.replaceAll(uiContext.pathSeparator, "/");
```
Comment (quockhanh) [inline svelte/src/lib/components/FileExplorer/Node.svelte:154]: Replace "/" by Node.ts->PATH_SEPARATOR  constant

However,  right here we have relatedSuiteFilePath  (absolute physical file path),  why do we parse to rel-path then pass into dialog,  but in the dialog, we parse from rel-path to  full-path again to modify test-suite file content?
Is it possible to pass the related-suites-full-path here into dialog?

File: svelte/src/lib/components/Application.svelte [hunk 11]
```diff
@@ -213,3 +215,5 @@
 {/if}
 
 <AboutDialog bind:showDialog={showAboutDialog} />
+
+<DeleteTestCaseDialog bind:showDialog={showDeleteTestCaseDialog} />
```
Comment (quockhanh) [inline svelte/src/lib/components/Application.svelte:219]: I dont think we need DeleteTestCaseDialog in Application.svelte  
(We already use it in FileExplorer/Node.svelte)

File: svelte/src/lib/components/FileExplorer/Node.svelte [hunk 12]
```diff
@@ -1,295 +1,338 @@
...
+    console.log("generateProjectMetadata", meta);
+    console.log("testCase", testCase);
+    console.log("relatedTestSuiteRelPaths", relatedTestSuiteRelPaths);
+    showDeleteTestCaseDialog = true;
+  };
+
+  const handleMenuDelete = async () => {
+    if (type === FileType.TCase) {
+      await handleDeleteTestCase();
+      return;
```
Comment (quockhanh) [inline svelte/src/lib/components/FileExplorer/Node.svelte:167]: **Important**: From my view point, we should NOT let DeleteTestCaseDialog handle everything, such as:  
- Delete test case
- Modify test suite
- Close tab
- .... (may be something else in future)

With above approach, the dialog needs to handle 'delete file' logic and related stuff, such as close tab (maybe, in future, we need to do extra work as well).   Moreover, if there is any update about delete-file logic, we have to update 2 places:  FileExplorer/Node.svelte and DeleteTestCaseDialog.svelte

I would recommend:
- If that is the case we need to handle (example: file type == TCase), then:    // call -> handleTestCaseDeletion
- Perform additional logic, such as generate code metadata then check whether test case is being used by test suite,
- If no,  execute normal deletion workflow as other scenario / file
- If yes, collect info (related suite full path, test case id), then display Delete Test Case (confirmation) dialog
- If user confirm NO:  do nothing, stop
- If user confirm YES:  on the dialog,  perform test suite modification first,  then dispatch custom event  (event can be:  'delete'  or 'delete-node' with the parameter is node path (or corresponding (test-case) fileSystemPath)
- Outside dialog, at the Node.svelte,  just handle that custom event then execute normal node deletion logic  (include other logic like close tab)
- In summary,  logic of delete file should leave or centralize at Node.svelte.     custom Delete...Dialog just handle its own special logic like modify  test-suite content

File: svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte (hunk 6 above)
Comment (vothaison) [reply svelte/src/lib/dialogs/DeleteTestCaseDialog.svelte:19]: Will be fixed when we move the "delete" logic to Node.svelte
